all: periodically save fastcache to filesystem

ehnuje opened the pull request:
## Proposed changes

- fastcache, in-memory cache structure used to store trie node entries, can be saved to the filesystem by calling `admin.saveTrieNodeToDisk`
- Now it can be saved automatically with the specified interval, without calling API
- `state.trie-cache-save-period` is used to specify the interval, default value is ~24 hour~ 0, which means it is disabled by default. At first it will be started after randomly chosen duration `x`, 
which ranges from `0.5 * state.trie-cache-save-period <= x < 1.0 * state.trie-cache-save-period`  

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

ehnuje commented on storage/statedb/cache.go:
```diff
@@ -77,7 +86,7 @@ func (cacheType TrieNodeCacheType) ToValid() TrieNodeCacheType {
 // NewTrieNodeCache creates one type of any supported trie node caches.
 // NOTE: It returns (nil, nil) when the cache type is CacheTypeLocal and its size is set to zero.
-func NewTrieNodeCache(config TrieNodeCacheConfig) (TrieNodeCache, error) {
+func NewTrieNodeCache(config *TrieNodeCacheConfig) (TrieNodeCache, error) {
```
Were there any reason that we used the cache config as a struct, not a pointer? @aidan-kwon @winnie-byun @KimKyungup

aidan-kwon replied:
I put "struct" to omit `nil` checking for `*TrieNodeCacheConfig` and its fields. We can change it to "pointer" if appropriate `nil` checking is added together.

ehnuje replied:
@aidan-kwon 73570247525c4bd99d8ae4cd323ac4460730f103 sets default value of TrieNodeCacheConfig if it is nil. I think it would be enough. How do you think?

aidan-kwon replied:
@ehnuje Thank you. I think it's okay. I am gonna add more checking logic in `NewTrieNodeCache` to avoid panic caused by developers' mistakes after this PR is merged.

winnie-byun reviewed (commented):
1. Were there any time delay when it is written?
2. What do you think of setting time instead of just 24h? (ex. Save to file on 4am) Huge txs are not usually occured at dawn in Korean time zone.
3. What happens when `SaveToFile()` is called while there is an ongoing save process? Ex. when the save period is 24 hours and save takes 40 hours. What do you think of checking if there is any ongoing save, before save?

ehnuje commented:
> 1. Were there any time delay when it is written?

It takes about 7 minutes to write 50GB of cache with `runtime.NumCPU` goroutines on c5.18xlarge machine. As we don't want that number of CPUs to be used for saving cache, I think it takes quite longer than 7 minutes on typical CN machine, if we adjust that number. This should be checked and I'll update the result here.

> 2. What do you think of setting time instead of just 24h? (ex. Save to file at 4a.m.) Huge txs are not usually occured at dawn in Korean time zone.

It looks like a good idea. We can consider this option. 👍 

> 3. What happens when `SaveToFile()` is called while there is an ongoing save process? Ex. when the save period is 24 hours and save takes 40 hours. What do you think of checking if there is any ongoing save, before a new save?

It will return an error since `SaveTrieNodeCacheToFile` checks if ongoing cache-saving is running or not. Please check the below codes. `savingTrieNodeCacheTriggered` flag severs this feature.

```
func (db *Database) SaveTrieNodeCacheToFile(filePath string) error {
	if db.trieNodeCache == nil {
		return errDisabledTrieNodeCache
	}
	if db.savingTrieNodeCacheTriggered {
		return errSavingTrieNodeCacheInProgress
	}
	db.savingTrieNodeCacheTriggered = true
	start := time.Now()
	go func() {
		logger.Info("start saving cache to file", "filePath", filePath)
		if err := db.trieNodeCache.SaveToFile(filePath, runtime.NumCPU()/2); err != nil {
			logger.Error("failed to save cache to file",
				"filePath", filePath, "elapsed", time.Since(start), "err", err)
		} else {
			logger.Info("successfully saved cache to file",
				"filePath", filePath, "elapsed", time.Since(start))
		}
		db.savingTrieNodeCacheTriggered = false
	}()
	return nil
}
```

KimKyungup commented on storage/statedb/database.go:
```diff
@@ -1153,6 +1153,21 @@ func (db *Database) SaveTrieNodeCacheToFile(filePath string) error {
+	for {
+		select {
+		case <-ticker.C:
+			db.SaveTrieNodeCacheToFile(c.FastCacheFileDir)
```
@ehnuje If the period is too short, this can call `SaveTrieNodeCacheToFile` again during saving.
So could you remove goroutine in `SaveTrieNodeCacheToFile` and change caller code of the `SaveTrieNodeCacheToDisk` API?

ehnuje replied:
Answered in the below comment, please check.

KimKyungup commented:
@ehnuje Could you check it is safe if the process is killed during saving the file?
In that case, Can the temp file being ignored and removed? Also can the past dump file being used?

ehnuje commented:
@KimKyungup 

> @ehnuje If the period is too short, this can call `SaveTrieNodeCacheToFile` again during saving.
> So could you remove goroutine in `SaveTrieNodeCacheToFile` to wait for it finish and change caller code of the `SaveTrieNodeCacheToDisk` API?

Updated. Please check https://github.com/klaytn/klaytn/pull/828/commits/027fc2fd1ef7020e6e415baeb17a94044f05dede

> @ehnuje Could you check it is safe if the process is killed during saving the file?
> In that case, Can the temp file being ignored and removed? Also can the past dump file being used?

It depends on the situation when the process is killed. Though I'm not sure what "safe" means here, if there is any problem with the saved cache in the designated directory, it will just ignore the file and create an empty fastcache, as you can see in the below code snippet. 

If it was writing the file to its temp directory and got killed, then the node will just use the past fastcache dump file, if available. However, if it was killed during removing old directory and renaming the temp directory, this can't be guaranteed. In this case, as I stated above, it fails to read the data and generates an empty fastcache.

All in all, in my opinion, there won't be an "unsafe" case, though we might have a "underperformed" one. 

```
func LoadFromFileOrNew(filePath string, maxBytes int) *Cache {
	c, err := load(filePath, maxBytes)
	if err == nil {
		return c
	}
	return New(maxBytes)
}
``` 
 
> And is it possible to prevent to call `SaveTrieNodeCacheToDisk` API during periodic saving?

Updated. Please check https://github.com/klaytn/klaytn/pull/828/commits/027fc2fd1ef7020e6e415baeb17a94044f05dede

KimKyungup commented:
> 2\. What do you think of setting time instead of just 24h? (ex. Save to file at 4a.m.) Huge txs are not usually occured at dawn in Korean time zone.

I think syncing all CN/PN dump moment is not good. It makes not good performance in that moment.
@ehnuje @winnie-byun

KimKyungup reviewed (dismissed):
LGTM.

@ehnuje 
However, I am worried that there will be no other impact when applying this PR to the cypress CN/PN.
What kind of tests can we do more?

ehnuje commented:
@KimKyungup 

> > 2. What do you think of setting time instead of just 24h? (ex. Save to file at 4a.m.) Huge txs are not usually occured at dawn in Korean time zone.
> 
> I think syncing all CN/PN dump moment is not good. It makes not good network performance in that moment.
> @ehnuje @winnie-byun

I think it is a kind of trade-off, so there won't be a perfect solution. Though the overall performance would be degraded, we can guarantee that the performance won't be decreased due to cache saving in other times. 

> LGTM.
> 
> @ehnuje
> However, I am worried that there will be no other impact when applying this PR to the cypress CN/PN.
> What kind of tests can we do more?

Now I've set up test networks to check the time spent for saving cache to files while processing transactions from locust. In the test, the period is set as 2 hour and only 1 CPU is used to save files. I'll update the result later.

aidan-kwon commented:
@ehnuje I am afraid that the function causes disk I/O congestion. As you know, cypress nodes occasionally experience heavy disk I/O when they process blocks (or APIs). Is there any way to mitigate the problem?

ehnuje commented:
> @ehnuje I am afraid that the function causes disk I/O congestion. As you know, cypress nodes occasionally experience heavy disk I/O when they process blocks (or APIs). Is there any way to mitigate the problem?

@aidan-kwon Are we exhausting gp2 SSD's credit in heavy I/O situation? If so, this can be a problem. If it's not, I think this can be different since it does not use LevelDB to dump fastcache.

ehnuje commented:
## Test Environment

- Tested on 2 types of machines -> m5.2xlarge / m5.4xlarge
- 4CN-4PN-4EN network
- total accounts = 5m / active accounts = 5m
- 4000RPS / transferSignedTx
- 1 goroutine for saving fastcache / 2 hours period

## Test Results
- m5.2xlarge used 20GB for fastcache and it takes about 1m 20s

```
INFO[01/04,07:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/04,07:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.713302171s
INFO[01/04,09:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/04,09:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.901638559s
INFO[01/04,11:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/04,11:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m17.000154976s
INFO[01/04,13:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/04,13:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.972977168s
INFO[01/04,15:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/04,15:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m17.006610683s
INFO[01/04,17:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/04,17:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.974241873s
INFO[01/04,19:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/04,19:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.72137253s
INFO[01/04,21:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/04,21:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.569882589s
INFO[01/04,23:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/04,23:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.560867423s
INFO[01/05,01:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/05,01:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.690116515s
INFO[01/05,03:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/05,03:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.885111395s
INFO[01/05,05:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/05,05:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.96158232s
INFO[01/05,07:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/05,07:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.965555615s
INFO[01/05,09:57:05 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[01/05,09:58:22 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m16.962204983s
```

- m5.4xlarge used 40GB for fastcache and it takes about 3m

```
INFO[12/30,13:51:44 +09] [47] Initialize local trie node cache (fastCache)  MaxMB=38173 FilePath=/home/ubuntu/klaytn/data/fastcache LoadedBytes=0 LoadedEntries=0
INFO[12/30,13:51:44 +09] [5] LocalCache is used for trie node cache, start saving cache to file periodically  dir=/home/ubuntu/klaytn/data/fastcache period=2h0m0s
INFO[12/30,15:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/30,15:51:59 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=15.502502106s
INFO[12/30,17:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/30,17:53:05 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=1m21.596695692s
INFO[12/30,19:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/30,19:54:12 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=2m28.643572796s
INFO[12/30,21:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/30,21:54:29 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=2m45.519309523s
INFO[12/30,23:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/30,23:54:30 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=2m45.729763004s
INFO[12/31,01:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/31,01:54:33 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=2m48.730106871s
INFO[12/31,03:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/31,03:54:41 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=2m56.962067791s
INFO[12/31,05:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/31,05:54:36 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=2m51.829402423s
INFO[12/31,07:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/31,07:54:33 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=2m49.342253414s
INFO[12/31,09:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/31,09:54:35 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=2m51.099987779s
INFO[12/31,11:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/31,11:54:44 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=2m59.804522846s
INFO[12/31,13:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
INFO[12/31,13:54:59 +09] [47] successfully saved cache to file          filePath=/home/ubuntu/klaytn/data/fastcache elapsed=3m15.46064516s
INFO[12/31,15:51:44 +09] [47] start saving cache to file                filePath=/home/ubuntu/klaytn/data/fastcache
```

- while writing the cache, block insert time increased about 20%

![screencapture-3-35-227-140-3000-d-KnJhY-iWz-klaytn-main-overview-2021-01-05-11_34_21](https://user-images.githubusercontent.com/2115430/103601356-d08de100-4f4c-11eb-8f01-0f2054b2e1e8.png)

ehnuje commented:
Dear reviewers, I've changed PR to use only 1 goroutine for saving fastcache, as it seems okay from the above test results.

ehnuje commented:
> > 2. What do you think of setting time instead of just 24h? (ex. Save to file at 4a.m.) Huge txs are not usually occured at dawn in Korean time zone.
> 
> I think syncing all CN/PN dump moment is not good. It makes not good network performance in that moment.
> @ehnuje @winnie-byun

@KimKyungup As it takes not that much time for saving the cache, we can consider the option suggested by @winnie-byun How do you think? We can select some time period ex. 02:00-04:00 AM and allocate time slot to each CCO, which are not identical. So Netmarble starts writing cache at 02:00, while Kakao starts at 03:00.

KimKyungup commented:
@ehnuje @winnie-byun 
I think we can consider the random saving period for the node. Let's talk the idea in offline.

ehnuje commented:
With 6ba824b951e4f3df2a315669bca6c5bbc0221af3, now periodic dump is disabled by default.

KimKyungup commented on storage/statedb/database.go:
```diff
@@ -1129,28 +1128,49 @@ func (db *Database) UpdateMetricNodes() {
+
+// DumpPeriodically atomically saves fast cache data to the given dir with the specified interval.
+func (db *Database) SaveCachePeriodically(c *TrieNodeCacheConfig, stopCh <-chan struct{}) {
+	ticker := time.NewTicker(c.FastCacheSavePeriod)
```
How about setting random value at the 1st only?

```suggestion
	ticker := time.NewTicker(c.FastCacheSavePeriod)
```

ehnuje replied:
@KimKyungup If you're worried about consistent performance degradation after the reboot due to round change, I have no objection. How long do you think it is proper for random period? (1hour? FastCacheSavePeriod?)

ehnuje replied:
Reflected by f9e2dd790d6589e8e10b7d690dbf634dee0cc774, please check @KimKyungup

KimKyungup replied:
@ehnuje Thanks. :)

ehnuje commented:
@winnie-byun @jeongkyun-oh @aidan-kwon Dear reviewers, could you please review this change?
